### ISSUE TITLE
Common Test framework

### ISSUE THREAD

**Issue by leakingtapan** [aws/aws-k8s-tester#48]
## Description
We have several open source projects that need to be tested with utilities to create/tear down cluster, setting up containers under test. However, we developed similar testing utilities in different places which makes it harder to maintain and share any improvements to the utility. Following are the list of testing utilities:
* EBS CSI - https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/master/hack/run-e2e-test
* EFS CSI - https://github.com/kubernetes-sigs/aws-efs-csi-driver/pull/55
* ALB - https://github.com/kubernetes-sigs/aws-alb-ingress-controller/blob/master/scripts/utils/cluster.sh

We are going to use this repo to share golang testing code that could be used by multiple open source projects including EBS/EFS/FSx CSI drivers, CNI plugins, App mesh webhook and Alb Ingress controller.

By hosting the code in aws-k8s-tester, we get several benefits:
* it's written in go, so that it can be imported by other go project easily
* it's centralized, so that we don't need to duplicate similar logic in each projects.

## Goals
* It is a go library. We will see if being a go library will cause any dependency complication during import. Otherwise, we will make it a cli binary
* It can create/tear down Kubernetes cluster
* It can support both EKS cluster and Kops cluster
* It can be configured with commands to build the project (eg, CSI driver) given `make` comand 
* It can be configured to build docker image given `docker build` command
* It can be configured to push to ECR
* It can be configured to deploy/remove container images (eg, CSI Driver)
* It can be configured to run test given test command `go test...`

## Non-Goals
* Deciding which test framework to use for running the test. It could be go test or ginkgo, which should be fully decided by end developer. 
* Deciding what test case to run. The test case is project specific. Only end developer knows what are the best use cases to run.

## Action Items:
- [x] rewrite EBS CSI Driver [`run-e2e-test`](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/master/hack/run-e2e-test) https://github.com/aws/aws-k8s-tester/issues/49 @leakingtapan @wongma7 
- [x] use the framework in FSx CSI Driver or Alb Ingress @leakingtapan 
- [x] Sending out PR review and merge https://github.com/tiffanyfay/aws-k8s-test-framework/pull/1 into tester @tiffanyfay

@wongma7 @gyuho @M00nF1sh @tiffanyfay  @nckturner @mogren

**Comment by mogren** [comment]
@leakingtapan Great initiative! The Kops compatibility is definitely interesting to me as well.

What is the next step? This repo already has a lot of stuff, and @tiffanyfay has a lot in this PR as well: https://github.com/tiffanyfay/aws-k8s-test-framework/pull/1

Should we split this up in a few tasks maybe? Adding @jaypipes to the discussion as well, since I know he has a lot of ideas around testing... 😀

**Comment by leakingtapan** [comment]
In my mind, one next step will be we rewrite the `run-e2e-test` script in EBS CSI driver here, and use it in one other project to get us some ideas of how it works. Creating some action items to keep them in track:

[ ] rewrite EBS CSI Driver [`run-e2e-test`](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/master/hack/run-e2e-test)
[ ] use the framework in FSx CSI Driver or Alb Ingress

**Comment by leakingtapan** [comment]
> What is the next step? This repo already has a lot of stuff, and @tiffanyfay has a lot in this PR as well: tiffanyfay/aws-k8s-test-framework#1

Since @tiffanyfay's PR are providing testing methods to interact with Kubernetes API objects, they are independent of the cluster creation, also has the potential to be used by other project. @tiffanyfay we can start merging your code here in parallel if you feel they are mature enough

**Comment by leakingtapan** [comment]
Test framework POC is done. Next step is try is out in a different project

**Comment by jaypipes** [comment]
Sorry for the delay in commenting on this. My initial reaction is that it's an excellent initiative, however the scope of what you want this repo to do is too large. In particular, I would advise removing the following items from the scope of what this repo is responsible for:

* It can be configured with commands to build the project (eg, CSI driver) given make comand
* It can be configured to build docker image given docker build command
* It can be configured to push to ECR
* It can be configured to deploy/remove container images (eg, CSI Driver)

All of the above items belong in a separate (probably private) repo that houses build scripts/tools. I would advise keeping this particular repo as just the test infrastructure and fixture setup.

**Comment by leakingtapan** [comment]
> All of the above items belong in a separate (probably private) repo that houses build scripts/tools. I would advise keeping this particular repo as just the test infrastructure and fixture setup.

Yep. The framework is not going to provide any utils to build project, build docker image, push to ECR, deploy the addon or run the test, because those are addon specific. The framework will provide a hook point to let addon develop to define those. :)

See how I am running e2e tests in EBS CSI driver using the framework: https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/master/tester/single-az-config.yaml

**Comment by leakingtapan** [comment]
close the issue, since this is done
